libcontainer/rootfs_linux.go

```diff
@@ -594,6 +596,40 @@ func createDevices(config *configs.Config) error {
 }
 
 func bindMountDeviceNode(dest string, node *configs.Device) error {
+	if st, stErr := os.Lstat(dest); stErr == nil {
+		// already exist (#2450)
+		if st.Mode()&os.ModeSymlink != 0 {
+			st, stErr = os.Stat(dest)
+			if stErr != nil {
+				return errors.Wrapf(stErr, "device node %q is already populated as a symlink but broken", dest)
+			}
+		}
+		if st.Mode()&os.ModeDevice == 0 {
+			return errors.Errorf("device node %q is already populated but not a valid device: %+v", dest, st.Mode())
+		}
+		switch node.Type {
+		case configs.BlockDevice:
+			if st.Mode()&os.ModeCharDevice != 0 {
+				return errors.Errorf("block device node %q is already populated but not a block device: %+v", dest, st.Mode())
+			}
+		case configs.CharDevice:
+			if st.Mode()&os.ModeCharDevice == 0 {
+				return errors.Errorf("char device node %q is already populated but not a char device: %+v", dest, st.Mode())
+			}
+		default:
+			// NOP
+		}
+		ust, ok := st.Sys().(*syscall.Stat_t)
+		if !ok {
+			return errors.Errorf("device node %q is already populated but not a valid device with *syscall.Stat_t: %+v", dest, st.Mode())
+		}
+		major := int64(unix.Major(ust.Rdev))
+		minor := int64(unix.Minor(ust.Rdev))
+		if major != node.Major || minor != node.Minor {
+			return errors.Errorf("device node %q is already populated but has unexpected number %d:%d (expected %d:%d)", dest, major, minor, node.Major, node.Minor)
+		}
+		return nil
+	}
```

Comment:
This code would be much better written something like this:

```suggestion
	var st unix.Stat_t
	if err := os.Lstat(dest, &st); err != nil {
		if !os.IsNotExist(err) {
			return errors.Wrapf(err, "could not stat existing path for device node: %q", dest)
		}
	} else {
		var gotType configs.DeviceType
		switch st.Mode() & os.ModeType {
		case os.ModeSymlink:
			// If the path is a symlink, we remove it.
			// NOTE: These semantics might be incorrect, but I'm not convinced
			//       that the semantics in the original PR are correct either.
			//       Resolving symlinks is a pretty dangerous game and I'd
			//       prefer not to do it if possible.
			if err := unix.Unlink(dest); err != nil {
				return errors.Wrapf(err, "device node %q already exists and is a symlink", dest)
			}
			goto create
		case os.ModeDevice:
			gotType = configs.BlockDevice
		case os.ModeDevice | os.ModeCharDevice:
			gotType = configs.CharDevice
		default:
			return errors.Errorf("device node %q already exists but not a valid device: %o", dest, st.Mode())
		}

		gotMajor := int64(unix.Major(st.Rdev))
		gotMinor := int64(unix.Minor(st.Rdev))
		if gotType != node.Type || gotMajor != node.Major || gotMinor != node.Minor {
			return errors.Errorf("device node %q already exists but incorrect type [%s %d:%d] (wanted [%s %d:%d])", dest, gotType, gotMajor, gotMinor, node.Type, node.Major, node.Minor)
		}
		return nil
	}
create:
```

Though I'm not sure what the right handling of symlinks is. Each option seems suboptimal to me...

Reply:
Can we error out if a symlink path has '/', otherwise keep the link?

Reply:
doesn't work on rootless
```
not ok 79 runc run [tty and ptmx in devices (#2450)]
# (in test file tests/integration/tty.bats, line 238)
#   `[ "$status" -eq 0 ]' failed
# runc list (status=0):
# ID          PID         STATUS      BUNDLE      CREATED     OWNER
# runc spec --rootless (status=0):
# 
# runc run test_busybox (status=1):
# time="2020-06-05T10:29:46Z" level=warning msg="exit status 1"
# time="2020-06-05T10:29:46Z" level=error msg="container_linux.go:353: starting container process caused: process_linux.go:450: container init caused: rootfs_linux.go:71: creating device nodes caused: device node \"/tmp/busyboxtest/rootfs/dev/tty\" already exists but not a valid device: 20666"
```

Reply:
modified to pass rootless